### PR TITLE
feat(servicecatalog): Create TagOptions Construct

### PR DESCRIPTION
**Pull request by arcrank**
Fixes:  [#17753](https://github.com/aws/aws-cdk/issues/17753)

Previously TagOptions were defined via an interface and we only created the underlying
resources upon an association.  This broke CX if tagoptions were mangaged centrally.  We move to
make the TagOptions class a wrapper around aggregate individual TagOptions.

BREAKING CHANGE: `TagOptions` now have `scope` and `props` argument in constructor, and data is now passed via a `allowedValueForTags` field in props

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts:92]: rename this back to `tagOptions`

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:29]: Once you make this class a Construct, `cdk.Stack.of(scope)` will become just `this`.

**Review by skinny85 — DISMISSED** [review]
First round of comments. We definitely need to minimize the amount of changes we are making to the tests in this PR, as we're changing production code at the same time. I also thing `TagOptions` should become a full Construct.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:63]: We're missing a unit test exercising the scenario outlined in https://github.com/aws/aws-cdk/issues/17753#issue-1066183886.

File: packages/@aws-cdk/aws-servicecatalog/README.md
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/README.md:204]: It turns out this was referenced from a bad example of what TagOptions are used for, so wanted to make it more inline with end usage.

File: packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts:170]: mainly just for logical grouping, I wanted the constraints to all be together.  Either that or just doing it alphabetically always, just want to have some consistency.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:14]: Okay, I had this before but removed it - the reason being is that there is not really an `id` since there isn't an actual umbrella resource, this just creates a list of ones automatically.  

Or, would this suggest that perhaps passing in the `id` to `TagOptions` I could then pass that along to each of the underlying ones so that it becomes part of its path?  I could see that being useful if you are ever making multiple ones and are looking at the end resources.

File: packages/@aws-cdk/aws-servicecatalog/test/integ.tag-option.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.tag-option.ts:41]: It felt appropriate to break it out since I moved the unit tests to their own file. That I felt was more appropriate since before we had input validation checks for tag options in the portfolio unit tests.  If you feel this is not necessary and instead can be accomplished in either of the existing integ tests I am fine with getting rid of it.

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: okay, that makes sense.  I'll remove changes to the individual unit test files and just have the added tests here.  In separate PR we can remove the redundancies.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:63]: will add.  Wasn't sure if it was appropriate since we don't do similar tests for other constructs.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:14]: You don't have to pass the `id` anywhere - if you create the underlying resources under `this`, that identifier will automatically become part of their unique ID (as it will be part of their path in the construct tree).

File: packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts:170]: I would leave it where it is now.

File: packages/@aws-cdk/aws-servicecatalog/README.md
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/README.md:204]: OK.

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: 👍

File: packages/@aws-cdk/aws-servicecatalog/test/integ.tag-option.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.tag-option.ts:41]: Since I already recommended against breaking out the unit tests to their own file in a different comment, I think we can also get rid of this.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:63]: I mean, it's literally the entire reason we're doing this PR. Seems like it deserves a unit test confirming it actually solves the reported problem 🙂.

**Comment by skinny85** [comment]
> This should address: #17753

Let's add a `Fixes #17753` to the PR description then.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:18]: sure, I kept it public at first since it seemed like people might find it useful, but without perhaps helper methods probably not very clear.

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts:316]: restored

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: As a follow up, I think we need to change the functionality and some of these tests, I should've explicitly put this in breaking changes.

For example, I don't think the idempotent associations were that meaningful, and maintaining it now becomes difficult since we would have to scan entire construct tree for the app to find if there is similar key-value, since they could be in different `TagOptions` constructs, I don't believe we can use `tryFindChild`.  I will continue to look for way around this, since this breaks current functionality.

Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: We could solve problem partially by storing the key and value alongside the CfnTagOptions.  And checking if those exist.

Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: And again the tests that have input validation are now not meaningful, previously it would break when calling the `associate` method.  But now it will fail to create the `TagOptions` themselves, thus no need to test it in the portfolio unit tests.

Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: Let's just remove the tests that don't make sense anymore.

Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: > But now it will fail to create the `TagOptions` themselves, thus no need to test it in the portfolio unit tests.

It doesn't matter where it fails. Like I said, we shouldn't be moving the tests methods at the same time we are changing production code.

So, please leave the validation tests where they are.

Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:274]: We can move them later, in a separate PR, that doesn't touch production code, only tests.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:112]: yup got it

File: packages/@aws-cdk/aws-servicecatalog/lib/product.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/lib/product.ts:137]: remove and use package.json fix instead

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts:334]: this test (and functionality) is no longer valid.

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts:334]: These are also not really valid but we can modify it slightly to still work, unfortunately still a lot of git churn.  Moving the creation away from the association means that early failures happen on creation not via association calls.

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:324]: same as above.

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:341]: can sort of be added back in?

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [inline packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: The cdk linting doesn't like this setup (no interface and no props struct) but I don't particularly see value of having classes that are just pass throughs.  I don't know if making it props with a `tagOptions` field is better since it seems redundant to do

```ts
const tagO = new servicecatalog.TagOptions(this, 'TagOptions, {
  tagOptions : {
    key1: ['v1', 'v2', 'v3'],
  }
});
```
Only if its a little bit of future proofing in case we want to add another field to props then it would be easier to make change.  
I will defer to you though on this.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: whoops I think I meant to say that the id is implicitly passed along. but yes you are correct.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: I think linting rules have become more stringent?  Build is failing since you apparently need to disable interface ref in js docs ( I had hoped I could just do it in package.json).

Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: Let's just introduce a `TagOptionsProps` interface as the third argument, like you suggested in https://github.com/aws/aws-cdk/pull/18314#discussion_r784495921.

Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: that works, leaves more wiggle room to not break things in future should changes happen and we want to modify props.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:88]: What if I do:

```
          key2: ['value2', 'value2', 'value2'],
```

? Seems like we are missing tests for these cases.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:101]: Can we name these something more descriptive than `stack1` and `stack2`? `tagOptionsStack` and `productStack`, for example?

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:126]: Missing empty line at the end of the file.

**Review by skinny85 — DISMISSED** [review]
We are getting closer, but I think this needs a couple more iterations before we can call it "Done".

File: packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts:170]: whoops, I interpreted where it is now to mean where it had been moved to.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:40]: sure,
the raw values - > `tagKey`: tagValue` and then the hashed value in map `tagName`

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts:334]: This is the interesting point that the original issue brought up, that I don't exactly know best way to handle this to fail faster than deploy time.  Tag Options are globally unique.  Moving towards the construct approach, then I don't you shouldn't be able to even make `tagOptions2`.  I know that we can check the constructs in a stack, but can we check all of the apps stack to see if same `tagOption` has already been made in an efficient manner?

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:15]: It can be confusing with keys in a map struct and keys to a tag, but I think we should stick with using tag key as the wording.  I don't know if I've seen tags referenced with a name and not just key-value.

File: packages/@aws-cdk/aws-servicecatalog/README.md
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/README.md:213]: Sure, I think they are case sensitive too but for docs we can use uppercase if that's the standard convention.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:44]: What about in the case of updating and modifying the values? e.g. we might have

`key: ['value1', 'value2', 'value3', value4']`

and then I update this to just be

`key: ['value1', 'value2', 'value4']`

I wouldn't expect to see 2 updated resources in template (1 deleted, 1 modified), only the one that was removed.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:45]: This is carryover from how we originally did things so that if you for example had duplicate values it would be idempotent and ignore.  I would prefer it if we could modify it so that creation is very strict, but that would change behavior again.

File: packages/@aws-cdk/aws-servicecatalog/test/integ.portfolio.expected.json
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.portfolio.expected.json:225]: can add it back, we don't have to provide it to handler since it's kind of a useless field.

File: packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/tag-option.test.ts:88]: I've added a test `tag options with duplicate values are idempotent`

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:17]: okay, improves naming of not having `tagOptions` field in `TagOptions`

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:35]: Because we iterate over the underlying values for creation and associating, if you do an empty one it will not fail and it won't create anything.  I can add a test for this case as well, or since this it's mostly useless we could validate and fail.

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts]: updated with modified wording and minimized changes.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts]: It's better for this to fail since it wouldn't align with other constructs to instantiate one with a certain (empty) configuration that wouldn't yield anything in template.  I've therefore added validation to make sure that the struct is not empty, and then that for each key the tag values are not empty, along with two tests for this.  There is not a max length for these lists since I believe that these can be altered by the customer.

**Review by skinny85 — COMMENTED** [review]
Looks great @arcrank. I basically have one last major question, and after that one is resolved, we can merge this one in.

File: packages/@aws-cdk/aws-servicecatalog/test/product.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/product.test.ts:361]: good call out.

File: packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts:144]: I have been able to maintain current behavior here.  I didn't try to maintain this before because from my OOP life it felt wrong to add a more coupling between the two resources, and I more or less tracked the impact of changes that would happen as all or nothing - since we could expect to see the TagOption resources changing I thought it might just go hand in hand to expect associations to change as well.

If this new version is fine, my understanding is now better for it.

File: packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/lib/tag-options.ts:61]: I can just make this a list, the map part was a relic of an earlier point where we considered adding lookup access to an underlying tagOption(s), and a map felt more efficient than searching list.

File: packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts:387]: I missed the earlier comment about extra lines, but in spirit of trying to keep changes to a minimum, I thought it was fine to keep it now since will be removing them in next PR immediately after this if that's okay.

**Review by skinny85 — APPROVED** [review]
Looks beautiful @arcrank, awesome job!

**Pull request merged**
